lib/cli/src/commands/publish.rs

```diff
@@ -0,0 +1,647 @@
+    };
+    Ok(SignArchiveResult::Ok {
+        public_key_id: personal_key.public_key_id,
+        signature: (minisign::sign(
+            Some(&minisign::PublicKey::from_base64(
+                &personal_key.public_key_value,
+            )?),
+            &private_key,
+            compressed_archive,
+            None,
+            None,
+        )?
+        .to_string()),
+    })
+}
+
+/// Opens an exclusive read/write connection to the database, creating it if it does not exist
+pub fn open_db() -> anyhow::Result<Connection> {
+    let db_path =
+        PartialWapmConfig::get_database_file_path().map_err(|e| anyhow::anyhow!("{e}"))?;
+    let mut conn = Connection::open_with_flags(
+        db_path,
+        OpenFlags::SQLITE_OPEN_CREATE
+            | OpenFlags::SQLITE_OPEN_READ_WRITE
+            | OpenFlags::SQLITE_OPEN_FULL_MUTEX,
+    )?;
+
+    apply_migrations(&mut conn)?;
+    Ok(conn)
+}
+
+/// Applies migrations to the database
+pub fn apply_migrations(conn: &mut Connection) -> anyhow::Result<()> {
+    let user_version = conn.pragma_query_value(None, "user_version", |val| val.get(0))?;
+    for data_version in user_version..CURRENT_DATA_VERSION {
+        log::debug!("Applying migration {}", data_version);
+        apply_migration(conn, data_version)?;
+    }
+    Ok(())
+}
+
+#[derive(Debug, Error)]
+enum MigrationError {
+    #[error(
+        "Critical internal error: the data version {0} is not handleded; current data version: {1}"
+    )]
+    MigrationNumberDoesNotExist(i32, i32),
+    #[error("Critical internal error: failed to commit trasaction migrating to data version {0}")]
+    CommitFailed(i32),
+    #[error("Critical internal error: transaction failed on migration number {0}: {1}")]
+    TransactionFailed(i32, String),
+}
+
+/// Applies migrations to the database and updates the `user_version` pragma.
+/// Every migration must leave the database in a valid state.
+fn apply_migration(conn: &mut Connection, migration_number: i32) -> Result<(), MigrationError> {
+    let tx = conn
+        .transaction_with_behavior(TransactionBehavior::Immediate)
+        .map_err(|e| MigrationError::TransactionFailed(migration_number, format!("{}", e)))?;
+    match migration_number {
+        0 => {
+            tx.execute_batch(include_str!("../../sql/migrations/0000.sql"))
+                .map_err(|e| {
+                    MigrationError::TransactionFailed(migration_number, format!("{}", e))
+                })?;
+        }
+        1 => {
+            tx.execute_batch(include_str!("../../sql/migrations/0001.sql"))
+                .map_err(|e| {
+                    MigrationError::TransactionFailed(migration_number, format!("{}", e))
+                })?;
+        }
+        2 => {
+            tx.execute_batch(include_str!("../../sql/migrations/0002.sql"))
+                .map_err(|e| {
+                    MigrationError::TransactionFailed(migration_number, format!("{}", e))
+                })?;
+        }
```

Comment:
Instead of writing a big `match` with lots of copy/paste, could we store this in an array constant at the top of the file?

```rust
const MIGRATIONS: &[(i32, &str)] = &[
  (0, include_str!("../../sql/migrations/0000.sql")),
  ...
];
```

Then, you'd do a `MIGRATIONS.iter().find_map(|(n, sql)| if n == migration_number { Some(sql) } else { None })` to find the right migration or handle errors appropriately.

Reply:
Yeah, just be aware that this part of the code is copy-pasted, not my original code. I just copied + refactored it for now.

Reply:
Fixed